### DISCUSSION TITLE
Odd failure mode (async button input spamming)

### DISCUSSION THREAD

**Discussion by GM-Script-Writer-62850** [micropython/micropython#15709] [General]
EDIT: so i managed to make it happen again, i think this this was a just a ID10T error, do you see a `btn.double.clear()` in my code cause i don't...
<hr>

It was probably between 1 and 2 AM and i was about to go to bed so the time i could put into trouble shooting was very limited

i went into the garage to unplug a fan and close some doors, unplugged the fan, closed some doors, and held my button to turn the light off and it went crazy

So i have this code, i held my button while the light was on (`GPIO.light.garage.off()` should be run)

instead it starts spamming double press, i have 4 momentary button in parallel (3 main wires, 1 is split at the end), i tried disconnecting the button I pressed, and it kept spamming away.
the light (120V AC LED light bulb; fused at 1A) is controlled via a solid state relay and a 2N2222A transistor

unplugged the pico and booting it back up and it was acting normal, any ideas as to what made it go crazy? never had it do this before, is has been in service for over a year

i can't think of a hardware level issue (condensation? emi pulse?) that would cause continuous spamming that would vanish with a power cycle

```
import uasyncio
import async_urequests as urequest
from primitives import EButton, ESwitch, WaitAny
EButton.long_press_ms=500
...

async def garage_buttons(pin):
	url="music/outputs/outputs.php?set=5&pico"
	btn=EButton(pin,suppress=True)#btn=EButton(pin,lopen=1)
	while True:
		btn.press.clear()
		btn.long.clear()
		event=await WaitAny((btn.press,btn.long,btn.double)).wait()
		if event in [ btn.press, btn.double ]:
			if event == btn.double:
				GPIO.light.garage.toggle()
			uasyncio.create_task(http(url))
		elif GPIO.light.garage.value():
			GPIO.light.garage.off()
		else:
			uasyncio.create_task(garageLight(90,1))

...
```

i also forgot i had a double press config till i was reading the code for this post...

**Comment by mendenm** [comment]
sounds like you don't have suitable pull-up turned on, and the input is floating?

**Reply by GM-Script-Writer-62850** [reply]
it is setup as a pull down, note there is a 1k safety resistor as well

```
logic_pins={
	"pullUp": [22, 26, 7],
	"pullDown": [15, 13, 12, 9, 8, 6] # Pin 15 is the relevant one
}

for x in logic_pins:
	i=0
	for y in logic_pins[x]:
		logic_pins[x][i]=Pin(y, Pin.IN, Pin.PULL_UP if x=="pullUp" else Pin.PULL_DOWN)
		i+=1

class GPIO:
	inputs={
		"back_door":Pin(12),
		"left_door":Pin(8),
		"right_door":Pin(9),
		"screen_door":Pin(13),
		"laundry_door":Pin(7),
		"living_door":Pin(26),
		"garage_button":Pin(15),# THIS PIN IS THE BUTTON 
		"living_button":Pin(22)
	}
...
```

maybe wen i unplugged the fan the emi flipped the pull down bit?

**Comment by mendenm** [comment]
have you tried measuring with a meter to see if it is really pulling hard to the rail?  This sounds an awful lot like floating.  Also, do you have the Schmitt trigger enabled?  A slow transition across threshold could cause this.  How long are the wires to the switch?  RF pickup?  You might need an external pull, maybe much lower resistance (few hundred ohms)  if there is interference.

**Reply by GM-Script-Writer-62850** [reply]
unless the pico has that built in and enabled by default no
it is just connected to 3.3v via a 1k resistor and a button

note that i have been using this setup for something like 2 years based on [the post i made on reddit](https://www.reddit.com/r/raspberrypipico/comments/101wx66/finally_got_one_of_my_3_boards_done/)

i did not pull out a meter, it was very late and i needed to get in bed this was Aug 16 at quarter to 1AM (found the spamming in my server logs)

**Reply by mendenm** [reply]
The RP2040 does have Schmitt triggers, and they are enabled by default, or should be, but maybe something is resetting that.  

Could there be a new, big source of RF interference which didn't used to be there?  If the wiring from the push button to the RP2040 is long, it makes an awfully nice RF antenna.  Note that the  electrostatic clamp diodes act as lovely RF detector diodes, so you may have built a crystal radio.  These MCUs are not designed to be hung on wiring running to various switches around a house.  The noise suppression on the input is designed to work on a circuit board.  This might be fixable with some software mitigation, but really you may need higher-voltage external circuits for the switches (24 VDC, e.g.) and solid state relays for the interface.  Houses are electrically quite noisy, and 3.3V logic was never meant for this.

**Reply by GM-Script-Writer-62850** [reply]
def not a _new_ one

definitionally rf antenna issues (used old telephone wire) i tried doing the software with interrupts and abandoned that cause of emi, i have not had any issues with my buttons/switches since using EButton/ESwitch, i have DHT22s that get knocked down now and i have them wires so i can power cycle them and get them back online

again this is the only instance so far in ~2 years for this issue

the relay i am using works with 5v on the low side and is rated at 2A on the high side

**Reply by GM-Script-Writer-62850** [reply]
so i managed to make it happen again, i think this this was a just a ID10T error, do you see a `btn.double.clear()` in my code cause i don't...

**Reply by davefes** [reply]
Sounds familiar ... https://github.com/orgs/micropython/discussions/12668
towards the bottom.
